File: July 2022/July_22.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode *s = nullptr, *g = nullptr, *st = nullptr, *gt = nullptr, *t = head;
        while(t != nullptr){
            if(t->val < x){
                if(s == nullptr)
                    st = s = t;
                else{
                    st->next = t;
                    st = st->next;
                }
            } else{
                if(g == nullptr)
                    gt = g = t;
                else{
                    gt->next = t;
                    gt = gt->next;
                }
            }
            t = t->next;
        }
        if(s == nullptr)
            return g;
        st->next = g;
        if(g != nullptr)
            gt->next = nullptr;
        return s;
    }
};

File: July 2022/July_22.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode smaller = null, greater = null, st = null, gt = null, tmp = head;
        while(tmp != null){
            if(tmp.val < x){
                if(smaller == null)
                    st = smaller = tmp;
                else{
                    st.next = tmp;
                    st = st.next;
                }
            } else{
                if(greater == null)
                    gt = greater = tmp;
                else{
                    gt.next = tmp;
                    gt = gt.next;
                }
            }
            tmp = tmp.next;
        }
        if(smaller == null)
            return greater;
        st.next = greater;
        if(greater != null)
            gt.next = null;
        return smaller;
    }
}
